cmd/storage-rest-server.go

```diff
@@ -0,0 +1,358 @@
+	volume := vars[storageRESTVolume]
+	filePath := vars[storageRESTFilePath]
+
+	buf := make([]byte, r.ContentLength)
+	_, err := io.ReadFull(r.Body, buf)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	err = s.storage.AppendFile(volume, filePath, buf)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+	}
+}
+
+// StatFileHandler - stat a file.
+func (s *storageRESTServer) StatFileHandler(w http.ResponseWriter, r *http.Request) {
+	if !s.IsValid(w, r) {
+		return
+	}
+	vars := mux.Vars(r)
+	volume := vars[storageRESTVolume]
+	filePath := vars[storageRESTFilePath]
+
+	info, err := s.storage.StatFile(volume, filePath)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	err = gob.NewEncoder(w).Encode(info)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+	}
+}
+
+// ReadAllHandler - read all the contents of a file.
+func (s *storageRESTServer) ReadAllHandler(w http.ResponseWriter, r *http.Request) {
+	if !s.IsValid(w, r) {
+		return
+	}
+	vars := mux.Vars(r)
+	volume := vars[storageRESTVolume]
+	filePath := vars[storageRESTFilePath]
+
+	buf, err := s.storage.ReadAll(volume, filePath)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	w.Header().Set("Content-Length", strconv.Itoa(len(buf)))
+	_, err = w.Write(buf)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+	}
+}
+
+// ReadFileHandler - read section of a file.
+func (s *storageRESTServer) ReadFileHandler(w http.ResponseWriter, r *http.Request) {
+	if !s.IsValid(w, r) {
+		return
+	}
+	vars := mux.Vars(r)
+	volume := vars[storageRESTVolume]
+	filePath := vars[storageRESTFilePath]
+	offset, err := strconv.Atoi(vars[storageRESTOffset])
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	length, err := strconv.Atoi(vars[storageRESTLength])
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	var verifier *BitrotVerifier
+	if vars[storageRESTBitrotAlgo] != "" {
+		hashStr := vars[storageRESTBitrotHash]
+		hash, err = hex.DecodeString(hashStr)
+		if err != nil {
+			s.writeErrorResponse(w, err)
+			return
+		}
+		verifier = NewBitrotVerifier(BitrotAlgorithmFromString(vars[storageRESTBitrotAlgo]), hash)
+	}
+	buf := make([]byte, length)
+	_, err = s.storage.ReadFile(volume, filePath, int64(offset), buf, verifier)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	_, err = w.Write(buf)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+	}
+}
+
+// ListDirHandler - list a directory.
+func (s *storageRESTServer) ListDirHandler(w http.ResponseWriter, r *http.Request) {
+	if !s.IsValid(w, r) {
+		return
+	}
+	vars := mux.Vars(r)
+	volume := vars[storageRESTVolume]
+	dirPath := vars[storageRESTDirPath]
+	count, err := strconv.Atoi(vars[storageRESTCount])
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	entries, err := s.storage.ListDir(volume, dirPath, count)
+	if err != nil {
+		s.writeErrorResponse(w, err)
+		return
+	}
+	err = gob.NewEncoder(w).Encode(&entries)
+	if err != nil {
```

Comment:
once writing to w we cannot write error again, it will throw multiple write calls error